arch/xtensa/src/esp32s2/esp32s2_oneshot.c

```diff
@@ -0,0 +1,462 @@
+
+  /* Call the callback */
+
+  oneshot->handler((void *)oneshot->arg);
+
+  /* Restore state */
+
+  oneshot->running = false;
+  oneshot->handler = NULL;
+  oneshot->arg = NULL;
+
+  /* Clear the Interrupt */
+
+  ESP32S2_TIM_ACKINT(oneshot->tim);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32s2_oneshot_initialize
+ *
+ * Description:
+ *   Initialize the oneshot timer wrapper.
+ *
+ * Input Parameters:
+ *   oneshot    Caller allocated instance of the oneshot state structure.
+ *   chan       Timer counter channel to be used.
+ *   resolution The required resolution of the timer in units of
+ *              microseconds.  NOTE that the range is restricted to the
+ *              range of uint16_t (excluding zero).
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success; a negated errno value is returned
+ *   on failure.
+ *
+ ****************************************************************************/
+
+int esp32s2_oneshot_initialize(struct esp32s2_oneshot_s *oneshot,
+                               int chan, uint16_t resolution)
+{
+  int ret = OK;
+
+  tmrinfo("chan=%d resolution=%d usecs\n", chan, resolution);
+
+  DEBUGASSERT(oneshot != NULL);
+  DEBUGASSERT(resolution > 0);
+
+  oneshot->chan = chan;
+
+  oneshot->tim = esp32s2_tim_init(chan);
+  if (oneshot->tim == NULL)
+    {
+      tmrerr("ERROR: Failed to allocate TIM %d\n", chan);
+      ret = -EBUSY;
+    }
+  else
+    {
+      uint16_t pre;
+
+      /* Initialize the remaining fields in the state structure. */
+
+      oneshot->running    = false;
+      oneshot->handler    = NULL;
+      oneshot->arg        = NULL;
+      oneshot->resolution = resolution;
+
+      /* Ensure timer is disabled.
+       * Change the prescaler divider with the timer enabled can lead to
+       * unpredictable results.
+       */
+
+      ESP32S2_TIM_STOP(oneshot->tim);
+
+      ESP32S2_TIM_CLK_SRC(oneshot->tim, ESP32S2_TIM_APB_CLK);
+
+      /* Calculate the suitable prescaler according to the current apb
+       * frequency to generate a period equals to resolution.
+       */
+
+      pre = (esp_clk_apb_freq() * resolution) / USEC_PER_SEC;
+
+      /* Configure TIMER prescaler */
+
+      ESP32S2_TIM_SETPRE(oneshot->tim, pre);
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: esp32s2_oneshot_max_delay
+ *
+ * Description:
+ *   Return the maximum delay supported by the timer.
+ *
+ * Input Parameters:
+ *   oneshot Caller allocated instance of the oneshot state structure. This
+ *           structure must have been previously initialized via a call to
+ *           esp32s2_oneshot_initialize();
+ *   usec    The maximum delay in us.
```

Comment:
```suggestion
 *
 * Output Parameters:
 *   usec    The maximum delay in us.
```
`usec`is an output parameter here.